### ISSUE TITLE
Feedback on Hermes support

### ISSUE THREAD

**Issue by kneth** [realm/realm-js#3940]
We [are working hard](https://github.com/realm/realm-js/pull/3792) on getting Hermes supported by Realm JavaScript, and we are releasing a series of pre-releases to test our progress.

We encourage you to try out the pre-release in your development environment (*don't use it in production yet*). Create [a new issue](https://github.com/realm/realm-js/issues/new?assignees=&labels=hermes&template=hermes.yml) with your observations, including details about iOS version, other JavaScript libraries. Please use the Hermes issue template.

We appreciate your feedback, but we cannot guarantee if and when we respond to your comment.

**Current state**
* Version: `v11.0.0-rc.1` (or `v11.0.0-rc.0` if you're using a React Native version between `0.66.0` and `0.68.2`)
* Both iOS and Android are supported
* Required React Native version is `0.69.0` or above 👈   this is very important, since JSI is not ABI stable and your app will simply crash if the version isn't correct. Use `v11.0.0-rc.0` for older versions of React Native).
* 0 of 276 tests are failing

**Installation**

We are keeping the pre-releases under the tag `hermes`, and you can install the latest pre-release using the following command:

```sh
npm install realm@hermes
```

For a more comprehensive list of TODOs, please see [our PR](https://github.com/realm/realm-js/pull/3792) and the source code for `TODO` comments.

**Comment by HSReact** [comment]
Great news thank you!!

**Comment by 8BallBomBom** [comment]
Wonderful news but as most of my development is mainly focused on Android... the waiting game continues.
Will jump in to test once that arrives 👍🏻

**Comment by cristianoccazinsp** [comment]
I'm also more interested on the Android side as non-hermes for iOS is quite fast and stable, but Android is the one that's slow and buggy.  However, great progress, I will be giving it a try as well on my free time.

What is the roadmap for Android? Does it need an entire new implementation as well, or is it the same code base with some tweaks?

**Comment by kraenhansen** [comment]
> What is the roadmap for Android? Does it need an entire new implementation as well, or is it the same code base with some tweaks?

We fully understand the need to get this out on Android ASAP too, and we're full-throttling into that now ..
The only reason we're pre-releasing for iOS and not Android at this moment is because we've hit less road-blocks in our implementation on iOS.

To be more detailed: We have a couple of outstanding issues in the way we bootstrap the module and the ability to catch exceptions natively on Android that we have to solve for a release to be usable. The good news is that majority of the code is shared between iOS and Android and we mainly need to solve issues in code and build-scripts for the part that bootstraps our .so library on Android.

**Comment by tedi** [comment]
Great news, thank you for all your hard work. We will be testing this in our next dev cycle and report back.

**Comment by moonpang** [comment]
Great new!!!

**Comment by joaodematejr** [comment]
I created a new application and installed it but it doesn't open the application, some Flipper log

![Screen Shot 2021-09-04 at 01 29 55](https://user-images.githubusercontent.com/17418041/132082412-f4b181a4-94f4-4143-9baf-127ca935de2f.png)

![Screen Shot 2021-09-04 at 01 29 32](https://user-images.githubusercontent.com/17418041/132082402-35ef247b-8458-4b69-83ee-acd0ffa2a9b0.png)

![Screen Shot 2021-09-04 at 01 18 07](https://user-images.githubusercontent.com/17418041/132082139-859b4b21-d6af-4fc7-b946-f49f216ac0aa.png)

![Screen Shot 2021-09-04 at 01 12 29](https://user-images.githubusercontent.com/17418041/132082034-668758fa-9120-4d51-ab86-80a17174de92.png)



Process:               realmHermes [16230]
Path:                  /Users/USER/Library/Developer/CoreSimulator/Devices/A0FFE31C-C19D-4CAC-917B-010B6CE3F4A2/data/Containers/Bundle/Application/732FAEC5-5CE9-4101-B41A-59C0DC19BF48/realmHermes.app/realmHermes
Identifier:            realmHermes
Version:               1.0 (1)
Code Type:             X86-64 (Native)
Parent Process:        launchd_sim [11811]
Responsible:           SimulatorTrampoline [7116]
User ID:               501

Date/Time:             2021-09-04 01:10:10.191 -0300
OS Version:            macOS 11.5.2 (20G95)
Report Version:        12
Anonymous UUID:        1E47EC22-9B28-370B-4B8C-FC36E91171B3


Time Awake Since Boot: 5600 seconds

System Integrity Protection: disabled

Crashed Thread:        12  com.facebook.react.JavaScript

Exception Type:        EXC_BAD_INSTRUCTION (SIGILL)
Exception Codes:       0x0000000000000001, 0x0000000000000000
Exception Note:        EXC_CORPSE_NOTIFY

Termination Signal:    Illegal instruction: 4
Termination Reason:    Namespace SIGNAL, Code 0x4
Terminating Process:   exc handler [16230]

Application Specific Information:
CoreSimulator 757.5 - Device: iPhone 8 (A0FFE31C-C19D-4CAC-917B-010B6CE3F4A2) - Runtime: iOS 14.5 (18E182) - DeviceType: iPhone 8

Thread 0:: Dispatch queue: com.apple.main-thread
0   com.apple.CoreFoundation      	0x00007fff2038b171 __CFRunLoopRun + 2789
1   com.apple.CoreFoundation      	0x00007fff2038a1a7 CFRunLoopRunSpecific + 567
2   com.apple.GraphicsServices    	0x00007fff2b874d85 GSEventRunModal + 139
3   com.apple.UIKitCore           	0x00007fff246c14df -[UIApplication _run] + 912
4   com.apple.UIKitCore           	0x00007fff246c639c UIApplicationMain + 101
5   org.reactjs.native.example.realmHermes	0x000000010ed664d0 main + 112 (main.m:7)
6   libdyld.dylib                 	0x00007fff2025abbd start + 1



EXC_BAD_INSTRUCTION

EXC_BAD_INSTRUCTION

Process:               realmHermes [21541]
Path:                  /Users/USER/Library/Developer/CoreSimulator/Devices/D5A0AFC6-F590-4E42-8B9D-867AD36CB43A/data/Containers/Bundle/Application/6AF8443A-4B97-4CCE-86AC-9F90A66740CA/realmHermes.app/realmHermes
Identifier:            realmHermes
Version:               1.0 (1)
Code Type:             X86-64 (Native)
Parent Process:        launchd_sim [18331]
Responsible:           SimulatorTrampoline [7116]
User ID:               501

Date/Time:             2021-09-04 01:27:10.133 -0300
OS Version:            macOS 11.5.2 (20G95)
Report Version:        12
Anonymous UUID:        1E47EC22-9B28-370B-4B8C-FC36E91171B3


Time Awake Since Boot: 6600 seconds

System Integrity Protection: disabled

Crashed Thread:        12  com.facebook.react.JavaScript

Exception Type:        EXC_BAD_INSTRUCTION (SIGILL)
Exception Codes:       0x0000000000000001, 0x0000000000000000
Exception Note:        EXC_CORPSE_NOTIFY

Termination Signal:    Illegal instruction: 4
Termination Reason:    Namespace SIGNAL, Code 0x4
Terminating Process:   exc handler [21541]

Application Specific Information:
CoreSimulator 757.5 - Device: iPhone 12 (D5A0AFC6-F590-4E42-8B9D-867AD36CB43A) - Runtime: iOS 14.5 (18E182) - DeviceType: iPhone 12

Thread 0:: Dispatch queue: com.apple.main-thread
0   com.apple.CFNetwork           	0x00007fff237203e4 0x7fff2353f000 + 1971172
1   com.apple.CoreFoundation      	0x00007fff20392513 _CFRelease + 244
2   com.apple.CFNetwork           	0x00007fff235ed2a9 0x7fff2353f000 + 713385
3   libobjc.A.dylib               	0x00007fff20178c42 object_cxxDestructFromClass(objc_object*, objc_class*) + 83
4   libobjc.A.dylib               	0x00007fff2018abee objc_destructInstance + 64
5   libobjc.A.dylib               	0x00007fff20190d88 -[NSObject dealloc] + 21
6   com.apple.CFNetwork           	0x00007fff236fff51 0x7fff2353f000 + 1838929
7   libobjc.A.dylib               	0x00007fff20191604 objc_object::sidetable_release(bool, bool) + 174
8   com.apple.CFNetwork           	0x00007fff237017c5 0x7fff2353f000 + 1845189
9   libobjc.A.dylib               	0x00007fff20191604 objc_object::sidetable_release(bool, bool) + 174
10  com.apple.CFNetwork           	0x00007fff2377bddb 0x7fff2353f000 + 2346459
11  com.apple.CFNetwork           	0x00007fff23556f4f 0x7fff2353f000 + 98127
12  libobjc.A.dylib               	0x00007fff20191604 objc_object::sidetable_release(bool, bool) + 174
13  libobjc.A.dylib               	0x00007fff20192dc7 AutoreleasePoolPage::releaseUntil(objc_object**) + 175
14  libobjc.A.dylib               	0x00007fff20192cb6 objc_autoreleasePoolPop + 185
15  com.apple.CoreFoundation      	0x00007fff20432036 _CFAutoreleasePoolPop + 22
16  com.apple.CoreFoundation      	0x00007fff2038fd5c __CFRunLoopPerCalloutARPEnd + 41
17  com.apple.CoreFoundation      	0x00007fff2038b1a2 __CFRunLoopRun + 2838
18  com.apple.CoreFoundation      	0x00007fff2038a1a7 CFRunLoopRunSpecific + 567
19  com.apple.GraphicsServices    	0x00007fff2b874d85 GSEventRunModal + 139
20  com.apple.UIKitCore           	0x00007fff246c14df -[UIApplication _run] + 912
21  com.apple.UIKitCore           	0x00007fff246c639c UIApplicationMain + 101
22  org.reactjs.native.example.realmHermes	0x0000000109f3e4d0 main + 112 (main.m:7)
23  libdyld.dylib                 	0x00007fff2025abbd start + 1

**Comment by kraenhansen** [comment]
@joaodematejr thanks a lot for taking this for a spin! Can I get you to create [a new issue with this feedback](https://github.com/realm/realm-js/issues/new?assignees=&labels=hermes&template=hermes.yml)? GitHub comments are really bad at threaded conversations ...

**Comment by liamjones** [comment]
@kneth Does this require a specific version of RN? I'd like to give this a try with one of our apps but that's still on 0.63.4, is that up-to-date enough?

**Comment by kneth** [comment]
@liamjones Our tests are using 0.64.2, and you might have to bump to 0.64 for it to work.

**Comment by kraenhansen** [comment]
We just pre-released [`realm` at v10.20.0-alpha.1](https://github.com/realm/realm-js/releases/tag/v10.20.0-alpha.1) with support for Hermes running on Android too! 🎉 

Please note that React Native must be at is 0.66.0-rc.3 .. this is very important, since JSI is not ABI stable and your app will simply crash if the version isn't correct.
Also hot reloading of app is working on iOS only (Android will be coming soon).

As an experiment, we're trying out the GitHub release discussion feature: So if you have any experiences (good or bad) with this pre-release, please share them over at https://github.com/realm/realm-js/discussions/3978

**Comment by Baira** [comment]
> 

Tested this extensively with a pool of users in one of my apps. No issues reported so far!

**Comment by HSReact** [comment]
Awesome news

**Comment by kraenhansen** [comment]
@Baira thank you so much for doing that!
This helps us build confidence in the release, which moves us one step closer to releasing this on our "latest" branch.

**Comment by cristianoccazinsp** [comment]
Where can I find currently known issues? 

Is it possible to test only for Android with this version? Which one is more stable, Android or iOS?

**Comment by cristianoccazinsp** [comment]
@kraenhansen I will be testing 10.20.0-alpha.2 on Android with Hermes over the next few days to about ~50 users.

I will also be testing 10.20.0-alpha.2 on iOS, but without Hermes (as it doesn't support INTL yet and polyfills are terrible), do you know if this Realm version without Hermes may be problematic? I've noticed you got rid of GCDWebServer dependency, so I'm afraid this version may have issues if not using Hermes.

**Comment by rajpootathar** [comment]
FPS drop significantly when app is offline the issue resolved in ` 10.10.1` but same issue happened in `10.20.0-alpha.2` Hermes.
I thinks the changes still not merged in this Hermes branch
 
 Related Issue: 
 https://github.com/realm/realm-js/issues/3882

**Comment by kraenhansen** [comment]
We have a release pending from our Hermes branch. It'll land first thing in the new year.

**Comment by YaoHuiJi** [comment]
> We have a release pending from our Hermes branch, which will pull in the latest improvements from `master`. It'll land first thing in the new year.

I almost cried when I heard that. Thanks to all the developers for your efforts. I had this problem when I was a little kid, and now I'm an happy old man.

**Comment by kraenhansen** [comment]
We're proud to announce that we've just released [Realm JS v10.20.0-beta.0](https://github.com/realm/realm-js/releases/tag/v10.20.0-beta.0) - the latest in the series of releases with support for the Hermes engine. With this release all our tests are now passing and we're even more confident in our approach.

See [issues with the `hermes` label](https://github.com/realm/realm-js/issues?q=is%3Aopen+is%3Aissue+label%3Ahermes) for outstanding issue that we'll tackle next. We would love to know more about your perceived runtime performance of this releases compared to your use-case requirements and experience from past releases. If you decide to take it for a spin, please report any bugs using [the hermes issue template](https://github.com/realm/realm-js/issues/new?assignees=&labels=hermes&template=hermes.yml) and tell us about your general experience via [this discussion on GitHub](https://github.com/realm/realm-js/discussions/4196).

Thanks for your patience and all your encouragement throughout this transition towards [Hermes as the default engine for React Native](https://reactnative.dev/blog/2021/10/26/toward-hermes-being-the-default)!

**Comment by liamjones** [comment]
@kraenhansen that's great to hear! Migration to Hermes is going to take us a bit of work due to other dependencies as well - does the beta still support the JSC as well so we can upgrade Realm first and then look at turning on Hermes afterwards?

**Comment by fronck** [comment]
@liamjones Yes, the beta still supports JSC.  You can use the `HERMES_ENABLED` environment variable to enable/disable JSI (Hermes).
See [here](https://github.com/realm/realm-js/blob/20e0cb89c5213b2d29346096d55a42bfcfe2249d/tests/ReactTestApp/ios/Podfile#L12) for example.

**Comment by elliottkember** [comment]
Fantastic news. I had just hit a conflict with react-native-reanimated, which seems to require Hermes, so this was perfect timing.

I switched this version in from 10.6.1 and have had no problems so far, but will report back if I find anything. Thank you for all the hard work!

**Comment by HSReact** [comment]
> Fantastic news. I had just hit a conflict with react-native-reanimated, which seems to require Hermes, so this was perfect timing.
> 
> I switched this version in from 10.6.1 and have had no problems so far, but will report back if I find anything. Thank you for all the hard work!

Thanks for sharing and do let us know how this goes, we are also using reanimated…though I believe it is reanimated 2 (which has far superior performance) which requires Hermes.

Either way we are happy to switch after hearing your success stories, do keep posting…

**Comment by YaoHuiJi** [comment]
I tried Realm JS v10.20.0-beta.0, but I met a terrible problem, sometimes I reopen my app, all data is missing, but when I check the default.realm file, all data is still in the file. It seems like sometimes my app don't read the `Documents/default.realm` file, but create a new db file in some place, because I can still create and update data when I found all data lost.

**Comment by kraenhansen** [comment]
@YaoHuiJi please [create an issue](https://github.com/realm/realm-js/issues/new?assignees=&labels=hermes&template=hermes.yml) with the relevant information allowing us to reproduce and eventually solve any bug there might be.

**Comment by YaoHuiJi** [comment]
> > I met a terrible problem
> 
> @YaoHuiJi please [create an issue](https://github.com/realm/realm-js/issues/new?assignees=&labels=hermes&template=hermes.yml) with the relevant information allowing us to reproduce and eventually solve any bug there might be.

Ok, because I tried it the day before yesterday, and now it has rolled back. I will try it again sometime in a few days.

**Comment by cristianoccazinsp** [comment]
> I tried Realm JS v10.20.0-beta.0, but I met a terrible problem, sometimes I reopen my app, all data is missing, but when I check the default.realm file, all data is still in the file. It seems like sometimes my app don't read the `Documents/default.realm` file, but create a new db file in some place, because I can still create and update data when I found all data lost.

Have you been able to reproduce it? Is it iOS or Android? Keep in mind iOS stores your data in a new absolute location with every new app update, so if for some reason there's an absolute path stored somewhere, it will be lost when you try to read it again.

**Comment by YaoHuiJi** [comment]
> > I tried Realm JS v10.20.0-beta.0, but I met a terrible problem, sometimes I reopen my app, all data is missing, but when I check the default.realm file, all data is still in the file. It seems like sometimes my app don't read the `Documents/default.realm` file, but create a new db file in some place, because I can still create and update data when I found all data lost.
> 
> Have you been able to reproduce it? Is it iOS or Android? Keep in mind iOS stores your data in a new absolute location with every new app update, so if for some reason there's an absolute path stored somewhere, it will be lost when you try to read it again.

Yes, I can reproduce it, but It only happened in the release version(iOS, real device), debug version works well.  what I found is weird...

1. Use Xcode to run my app(release) on my phone,  everything works well
2. Click 'Stop' in XCode to detach the app process
3. Reopen the app by tapping the app icon on iPhone
4. All data "disappeared", but I can create new data records in my App
5. If I repeat step 1, all data shows up again. Then repeat steps 2, 3, all data disappeared again.

When I check data in `Documents/default.realm`, all data exists, including the old data and the new data, the app didn't create another realm data file(as I guessed before), just the same one. The app can write data but read nothing, there is no crash, just no data can be read and showed up.  It only happened on the release version without Xcode attached, so I don't know how to debug it further...

hi @kraenhansen , I don't know if it's enough to create an issue, because I don't how to get more information.

packages related:
"react-native": "0.66.0",
"realm": "10.20.0-beta.0",
"redux": "^4.1.1",
"react-redux": "^7.2.4",

**Comment by zoobibackups** [comment]
> Fantastic news. I had just hit a conflict with react-native-reanimated (edit: v2), which seems to require Hermes, so this was perfect timing.
> 
> I switched this version in from 10.6.1 and have had no problems so far, but will report back if I find anything. Thank you for all the hard work!

Which specific version of realm and reanimated you are using, cs i still facing the same issue with reanimated v2.2.2.

**Comment by cristianoccazinsp** [comment]
Trying to upgrade to `10.20.0-beta.1` but it won't even let me start the app on iOS (`10.20.0-alpha.2` worked fine). Note that I'm not using Hermes on iOS (`:hermes_enabled => false`) but I am using it on Android.

I'm getting `Properties can only be defined on Objects` from `realm/lib/extensions.js` on line 364:
```
Object.defineProperties(
      realmConstructor.App.Sync.Subscriptions,
      getOwnPropertyDescriptors(subscriptionsMethods.static),
    );
```

**Comment by kraenhansen** [comment]
@cristianoccazinsp that's surely annoying, I've created a new issue for you, please follow that for progress.

**Comment by cristianoccazinsp** [comment]
@kraenhansen thanks, I'm assuming Both Hermes and JSI will be supported for iOS once Hermes support is fully released? Turns out Hermes is only useful for Android, but it's still lacking for iOS, so we don't have plans of upgrading to Hermes on iOS until a bunch of issues are resolved.

On the other hand, while using `10.20.0-alpha.2`, we started getting errors from looping through a realm's filtered collection. I don't have a lot of details yet, but at some point all code running queries start failing with the following error: `undefined is not an object (evaluating 'v.next')`. I'm sorry I don't have a stack trace, the the user that sent this log basically got this error from everywhere a realm query was looped over, and ultimately ended up in a hard crash: `EXC_BAD_ACCESS: Attempted to dereference garbage pointer 0x1c.`

I will open an issue for this if I see more logs coming through, but wanted to mention it on the hermes thread.

**Comment by HSReact** [comment]
We’re waiting for it to be a little more stable before we start using it, but use it we will…so happy to see the progress

**Comment by zoobibackups** [comment]
> @cristianoccazinsp that's surely annoying, I've created a new issue for you, please follow that for progress.

Please any tell me, which version of realm is working fine with reanimated v2 and herms enabled?????

**Comment by kraenhansen** [comment]
@zoobibackups 10.20.0-beta.1, installation instructions are in the topmost comment on this issue 👆

**Comment by mTondenier** [comment]
Hello everyone , i just installed the latest realm 10.20beta1 and with this version sometimes when i call the method objects() it returns me a strange [[Circular]] object. I don't think it's the expected behavior

**Comment by tomduncalf** [comment]
Hi @mTondenier, thanks for the report! Do you have an example or sample code you can share to reproduce this issue please?

**Comment by mTondenier** [comment]
Hi @tomduncalf  i can share some parts of the code : 

The schema : 

```
export const ExerciseSummarySchema = {
     name: EXERCISE_SUMMARY_SCHEMA,
     primaryKey: 'id',
     properties: {
          id: 'string',
          name: 'string',
          tags: 'string',
          functionnalGroups: 'string',
          difficulty: 'int',
          equipment: 'string',
          advice: 'string?',
          alias: 'string?',
          description: 'string?',
          videoAuthor: 'string?',
          authorInstagram: 'string?',
     },
};
```

functionnalGroups, equipement and tags are objects that are stringified , i don't know if it can help.

The query : 

 ```
const exerciseList = realm.objects(EXERCISE_SUMMARY_SCHEMA);
 console.log(exerciseList);
 console.log(exerciseList[0]);
```

the first console.log return : `[[Circular]]`
The second one surprisingly return the expected first item of the array

I'm new in js and react native but i don't think the error come from my code because it works if i downgrade realm.

**Comment by tomduncalf** [comment]
Thanks @mTondenier. 

I created a test app based on our [React Native "task list" template](https://github.com/realm/realm-js/tree/master/templates/react-native-template-realm-ts), enabled Hermes and [added a `console.log`, logging out the tasks](https://github.com/tomduncalf/issue-3940/blob/main/App.tsx#L58), and I do not see the `[[Circular]]` issue when logging them – I see the expected 
```[{"_id": [ObjectId], "description": "abc", "isComplete": false}, {"_id": [ObjectId], "description": "def", "isComplete": false}, {"_id": [ObjectId], "description": "ghi", "isComplete": false}, {"_id": [ObjectId], "description": "jkl", "isComplete": false}]```

I wonder if there's other stuff in your app which could be causing this problem? While my test app has a different schema to yours, I think that they are roughly equivalent in that they are both simple schemas with fields with primitive types, no references, etc: https://github.com/tomduncalf/issue-3940/blob/main/app/models/Task.ts#L17-L25

If you can think of anything else that might be different, or if you'd like to share your code in private, please let me know.

If you'd like to try out my test app to verify that it works for you, you should just be able to clone https://github.com/tomduncalf/issue-3940, then `npm install`, `npx pod-install` and `npm run ios`.

I will speak to the team to see if they have any ideas what else could be causing the `[[Circular]]` and will let you know if we think of anything.

**Comment by mTondenier** [comment]
Thanks to you @tomduncalf for anwsering so fast.

I just forgot to mention that the problem is only on android.

I will try to solve the problem and if i find where it come from i will tell you.

**Comment by tomduncalf** [comment]
Ah I only tested on iOS so that could be part of it! I’ll test android later today and let you know how it goes

**Comment by tomduncalf** [comment]
@mTondenier Could you please (create a new bug report issue)[https://github.com/realm/realm-js/issues/new?assignees=&labels=T-Bug&template=bug.yml] with the details of your setup and the issue you are having, and I will assist you with it there? Just so we can keep this main Hermes thread on topic. Thanks!

**Comment by kraenhansen** [comment]
Again, we're proud to announce that we've just released [Realm JS v10.20.0-beta.4](https://github.com/realm/realm-js/releases/tag/v10.20.0-beta.4) and [Realm JS v10.20.0-beta.5](https://github.com/realm/realm-js/releases/tag/v10.20.0-beta.5) - the latest in the series of releases with support for the Hermes engine. This is mostly a release rebased onto our recent [Realm JS v10.15.0](https://github.com/realm/realm-js/releases/tag/v10.15.0) and [Realm JS v10.15.0](https://github.com/realm/realm-js/releases/tag/v10.16.0) releases respectively. Although v10.20.0-beta.4 came with two minor bug-fixes as well.

Our current plan is to release the `hermes` branch as v11.0.0 together with a couple of other breaking changes around summer this year. Hopefully that will give us time to improve it further and the community the time to migrate to hermes (which is technically not a requirement, but will provide a much better debugging experience - as we're removing our support for the legacy chrome debugger).

**Comment by razmjoo-dev** [comment]
Will there be a problem if I use realm with hermes in production mode ?

**Comment by jaltin** [comment]
Hi,

Do you have any idea of when the Hermes version might be released? @kraenhansen you mentioned in your post from 13/4 that you planned to release it around the summer.

Asking as I am curious if you have the same plan still (it really affects our planning).

**Comment by HSReact** [comment]
> Hi,
> 
> Do you have any idea of when the Hermes version might be released? @kraenhansen you mentioned in your post from 13/4 that you planned to release it around the summer.
> 
> Asking as I am curious if you have the same plan still (it really affects our planning).

Yes, same for us...we are waiting for this to happen

**Comment by tomduncalf** [comment]
Hey @jaltin @HSReact, sorry the delay on this. We are investigating some test failures on the Hermes branch, trying to determine if they are actual issues or not. We hope to be able to investigate this this week and if we can fix the issue or clarify what is happening, we will release soon after. I will keep you updated. Thanks for your patience!

**Comment by tomduncalf** [comment]
Sorry @jaltin @HSReact, I just realised I misunderstood your question, I thought you were asking when we would have a new version of the Hermes branch, but I think you're actually asking when it will become the `latest` (default) version of Realm on npm?

Our plan is to wait until Meta make Hermes the default engine for React Native, and until they are recommending Flipper as the default way to debug React Native apps. Until this is the case, we feel that we can't push users to use our Hermes branch, especially as they'll need to use Flipper to debug.

**Comment by 8BallBomBom** [comment]
Seems like Hermes is going to become default pretty soon, starting with one of the upcoming releases.
Another sort of related question is if Realm has any plans to adopt the new architecture?

**Comment by HSReact** [comment]
> Sorry @jaltin @HSReact, I just realised I misunderstood your question, I thought you were asking when we would have a new version of the Hermes branch, but I think you're actually asking when it will become the `latest` (default) version of Realm on npm?
> 
> Our current plan (which is of course subject to change) is to wait until Meta make Hermes the default engine for React Native, and until they are recommending Flipper as the default way to debug React Native apps. Until this is the case, we feel that we can't push users to use our Hermes branch, especially as they'll need to use Flipper to debug.

Hermes will become the default.  We're all suffering slow performance, load times and the inability to exploit any new functionality in RN.

My biggest concern @tomduncalf is whether the Hermes support will make Realm access slow. What has been the feedback from beta testers who are using Realm Hermes with RN? 

Thanks for your hard work btw.

**Comment by tomduncalf** [comment]
@Pingou I would expect the next version to be a release candidate. We’d appreciate any feedback from our users on the production readiness of it. Obviously we can’t promise anything with regard to timelines on releasing this as there are many factors that can affect timings, but we will keep you updated.

In terms of production readiness, we have users who are using our Hermes branch in production and the only major issue we are currently aware of is a performance regression for both Hermes and non-Hermes apps, which we are actively investigating. This is only likely to be an issue if you are dealing with large amounts of reads/updates. You can find out more details in https://github.com/realm/realm-js/issues/4443.

I should also mention that our Hermes branch introduces a tighter coupling to a specific React Native version, as we depend on the JSI ABI which can change from RN release to release. We are investigating if there are alternatives to work around this, but for now this means you should use the RN version we recommend with Hermes.

**Comment by tomduncalf** [comment]
@HSReact as I mentioned above, we are aware of some performance regressions with our Hermes branch, see #4443 for more information. Whether these will be an issue for you or not depends on your workload, it's hard to say if it will have a real world impact without trying it out in your app – I'd expect apps with large volumes of data/reads/writes to be more affected.

In terms of resolving the issues, we believe that an upcoming RN release will fix the performance regression when Hermes is enabled – currently RN internally times every C++ <-> JS call which adds considerable overhead to Realm, but Meta have informed us that this timer will be disabled by default in the next version of Hermes. We are actively looking into ways to fix the performance regression when Hermes is disabled, which occurs for a different reason.

**Comment by tomduncalf** [comment]
@8BallBomBom regarding "Another sort of related question is if Realm has any plans to adopt the new architecture?" – I'd expect Realm to work with the new architecture enabled. If you're hitting issues here please let us know. Are there other specific aspects you are looking for us to support?

**Comment by 8BallBomBom** [comment]
Ahhh great to know, wasn't exactly sure about that one, will get testing once RN 0.69 drops 👍🏻 
Unsure really if there is anything else, once the next RN drops i'll be focused on using Hermes and Fabric.
If any issues arise then i'll definitely sprint to report them 😃

**Comment by HSReact** [comment]
> @HSReact as I mentioned above, we are aware of some performance regressions with our Hermes branch, see #4443 for more information. Whether these will be an issue for you or not depends on your workload, it's hard to say if it will have a real world impact without trying it out in your app – I'd expect apps with large volumes of data/reads/writes to be more affected.
> 
> In terms of resolving the issues, we believe that an upcoming RN release will fix the performance regression when Hermes is enabled – currently RN internally times every C++ <-> JS call which adds considerable overhead to Realm, but Meta have informed us that this timer will be disabled by default in the next version of Hermes. We are actively looking into ways to fix the performance regression when Hermes is disabled, which occurs for a different reason.

Great thank you very much for this, will check out #4443

**Comment by tradebulls** [comment]
@tomduncalf After reading the entire above thread and #4443, we are curious to know more about Android performance fixes. I would definitely try out the solution given by you in #4443  thread of disabling the timer in Hermes.cpp. Could you please help us with any solution for Android until a new release candidate of realm is not rolled out?

**Comment by tomduncalf** [comment]
Hey @tradebulls, I was focussing on iOS, but I was able to build Hermes for Android with the timer disabled. I hit some issues so I'm not 100% sure if this worked. I figured these steps out from their [CircleCI workflow](https://github.com/tomduncalf/hermes/blob/td/timer-disabled/.circleci/config.yml) and have reconstructed the steps from memory/my shell history, so apologies if I miss something!

1. Clone my fork of the Hermes repo from https://github.com/tomduncalf/hermes
2. Check out the `td/timer-disabled` branch, where I've cherry picked the commit from Hermes which disables the timer onto the v0.9.0 tag (which is what React Native needs)
3. Create a directory at the same level as your `hermes` clone called `hermes_ws` and `cd hermes_ws`
4. Set an env var pointing to this dir: `export HERMES_WS_DIR=\`pwd\``
5. Create a symlink to `hermes` inside `hermes_ws`: `ln -s ../hermes`
6. Configure the hermes compiler build: `hermes/utils/build/configure.py ./build`
7. Build the hermes compiler: `cmake --build ./build --target hermesc`
8. Build the hermes package: 
    ```
    export ANDROID_SDK="$ANDROID_HOME"
    export ANDROID_NDK="$ANDROID_HOME/ndk-bundle"
    cd "$HERMES_WS_DIR/hermes/android" && ./gradlew githubRelease
    ```
9. Unzip the build artefact, `hermes_ws/build_android/distributions/hermes-runtime-android-v0.9.0.tar.gz`
10. In your React Native project directory, overwrite the files in `node_modules/hermes-engine/android/` with the files you unzipped in step 9.

Obviously this isn't ideal and might be a bit tricky to integrate into your CI, but if you do try it out please let me know how you get on.

**Comment by tradebulls** [comment]
If the fix is available in latest RN 0.69.0 then we might not have to do this workaround.

**Comment by yadigar-prospr** [comment]
It looks like version #11.0.0-rc.0 perfectly runs on Android with Hermes also enabled with iOS didn't enable flipper yet 🥳

**Comment by kraenhansen** [comment]
@yadigar-prospr you beat me to it 😉 

We're super proud to announce that we've just released [Realm JS v11.0.0-rc.0](https://github.com/realm/realm-js/releases/tag/v11.0.0-rc.0) and [Realm JS v11.0.0-rc.1](https://github.com/realm/realm-js/releases/tag/v11.0.0-rc.1) - the latest in the series of releases with support for the Hermes engine. The first has support for React Native `0.66.0` through `0.68.2`, where-as the latter has support for the latest React Native `0.69.0` and `0.69.1`.

The most breaking part of this release is the official removal of our support for the legacy Chrome debugger and the need to `extend Realm.Object` when declaring your schema using "class-based models".

_Note_: We're in the process of writing a blogpost, outlining the work that has gone into making this a reality and hope to share some light on the decisions and tradeoffs we've had to make in the process.

_Also note_: We're entertaining the possibility of [deferring compilation](https://github.com/realm/realm-js/pull/4027) to your machines, to avoid this hard dependency on specific React Native versions.

Please take it for a spin! We love getting your feedback on the stuff we're making!

**Comment by cristianoccazinsp** [comment]
@kraenhansen does this mean we need to use rc.0 if we are using RN 0.68 and rc.1 if using RN 0.69?

I'm a bit nervous with this new React-Native architecture being the only available one in the near future, forcing everyone to upgrade to 0.69 (and any related library) and the new turbomodules implementation.

**Comment by kraenhansen** [comment]
> does this mean we need to use rc.0 if we are using RN 0.68 and rc.1 if using RN 0.69?

Yep - thats correct. I realise now that I could have signalled this even more using our peer dependency on React Native (but in all honesty I forgot to update it).

> I'm a bit nervous with this new React-Native architecture being the only available one in the near future.

I get you, it's annoying to have some dependencies that might be incompatible with the new architecture and some that aren't. To my knowledge it's been designed for gradual adoption, and it should be possible for app creators to mix between "old" libraries and newer libraries taking advantage of the new architecture.

We're in the unfortunate situation that we're shipping prebuilt binaries compiled against the JSI header, which introduce hard dependencies on React Native versions when this updates, but I believe that is far from the "normal" amongst libraries out there.

Please correct me if I'm wrong on any of those assumptions ☝️

**Comment by cristianoccazinsp** [comment]
@kraenhansen thanks for the answer.  My question is still more about the future versions once "RC" is made an official stable vesion. Do you guys plan to support both RN 0.68 and 0.69 for the near future, or is the plan to drop support to the "old architecture" and only release versions for RN >= 0.69 + turbomodules?

Thanks!

**Comment by cz2013** [comment]
@kraenhansen Will the realm team continue to maintain the rc.0 version? If there is a bug in using rc.0, will it continue to upgrade the release version, or will only the rc.1 version be maintained in the future? This is very important whether we use rc.0 or not. Thanks!

**Comment by tomduncalf** [comment]
Hi both, I'll speak to Kræn to confirm what our plans are here and get back to you once I have clarification

**Comment by renjuthoovara** [comment]
@kneth @kraenhansen Could you please let me know When we can expect the production release to support Hermes? We are facing the same issues so we can take a decision based on this

**Comment by mordechaim** [comment]
Does rc.1 work on react-native@0.70.0-rc.4? 

I've spent my entire day to get it to work but I'm running into `RealmProvider` not being available from @realm/react

**Comment by takameyer** [comment]
> I'm running into RealmProvider not being available from @realm/react
The `RealmProvider` is returned in the object returned by `createRealmContext`.  This allows one to create and use multiple Realms.
https://github.com/realm/realm-js/tree/master/packages/realm-react#setting-things-up

**Comment by HSReact** [comment]
An update on the current status from the Realm team would be nice @kraenhansen @kneth Thank you

**Comment by osama-kwanso** [comment]
Hi @kneth Its been an year now since last update. And can i use this @hermes version in production now?

**Comment by osama-kwanso** [comment]
> An update on the current status from the Realm team would be nice @kraenhansen @kneth Thank you

Yes its been a year since they last updated. @kneth we really wanna use this in production so please guide use can we now?

**Comment by HSReact** [comment]
From your outline:

"We cannot promise any specific release date, only that we will release it when we and the code are ready."

The writing is on the wall for me.  You guys are struggling to make this happen and this does not bode well.  Either you don't have the resources, or the commitment. 

I mean, you need to make this work now. As paying customers who have stayed with you, this is the least we can expect.

Sad News for me.

**Comment by Baira** [comment]
I've been following this thread for a long time. I guess a lot of people also do silently. I have many projects with realm already and everytime we start a new one, realm seems to be holding us back.

**Comment by mordechaim** [comment]
> The RealmProvider is returned in the object returned by createRealmContext. This allows one to create and use multiple Realms.

Yes, I'm aware of this, I'm trying to migrate an existing app from 0.68.2 to 0.70.0-rc.4.

I updated my dependency from realm@11.0.0-rc.0 to 11.0.0-rc.1 but the following code

```js
const { RealmProvider, useRealm, useObject, useQuery } = createRealmContext(myRealmConfig);
```
Will fail with
```
cannot read property RealmProvider from undefined
```

Now, I'm still having trouble building on the new architecture so I was trying with `newArchEnabled=false`, this could be the issue, although it did work perfectly on 0.68 with realm@11.0.0-rc.0

**Comment by takameyer** [comment]
@mordechaim can you please create a new issue for this?  We haven't officially made an update for 0.70.0-rc.4.  It is possible that we need to create a new release for this. Since the Hermes migration, we are hard bound to any updates to `react-native` and need to create a new release if there are any changes to the Hermes engine that are not compatible with the previous version.

**Comment by ianpward** [comment]
Hi All, Product for Realm here. As the engineering team mentioned, we have some outstanding work that we want to wrap up before moving v11 to master. Kenneth mentioned a performance regression that we uncovered and right now, that performance regression seems to be due to React Native's underlying architecture - so unfortunately, we are waiting for them to fix this issue. Once they fix this, we will need to retest to see where we are at. 

However, there is nothing stopping you from using our RC version in production today. Only you, as the app developer and owner, can make the call whether you are comfortable with using a release candidate in production. In fact, you already are if you are using React Native as a platform, which is pre-1.0 framework with an ever-changing and breaking API. This makes it difficult for us to keep up without making breaking changes in our version support. For an embedder such as ourselves, we've had to fight quite a bit against the framework and overwrite private APIs in order for us to even work, which is likely why you don't see many native C++ libraries that need to reflect across the bridge. We hope that the new architecture with  TurboModules and Fabric will make this easier and easier in the future. 

I would advise you to first run all your unit and performance testing with our RC release and if this passes, along with a successful UAT - you can then make a call whether you feel comfortable releasing to the app store. 

All this being said, we are in contact with the React Native team, and are hopeful that fixes will come soon - I hope that we will release our v11 master before the end of the year. If there is anyone who would like to discuss their situation further I am always available for a chat and you can reach out to me at ian.ward@mongodb.com and we can setup a meeting time.

**Comment by martinpoulsen** [comment]
Excited to see progress on v11! However, I'm still confused about the compatibility:

1. Will it require some minimum/specific RN version?
2. Is enabling Hermes required?
3. Is enabling the new architecture required?

**Comment by kneth** [comment]
> Will it require some minimum/specific RN version?

Yes, but we haven't decided which one yet. [0.70](https://reactnative.dev/blog/2022/07/08/hermes-as-the-default) is just around the corner, 

> Is enabling Hermes required?

With our JSI implementation, we are able to support both JSC and Hermes. When we release v11, we will try to document the advantages and disadvantages for both so you can make an informed decision.

> Is enabling the new architecture required?

Currently our focus is on JSI. TurboModules will be worth investigating at some point.

**Comment by martinpoulsen** [comment]
> Currently our focus is on JSI. TurboModules will be worth investigating at some point.

Does this mean that enabling the new architecture will not be an option in the foreseeable future when depending on Realm?

**Comment by HSReact** [comment]
So, can I ask a straight question.

I would like to use Hermes to get the benefits (faster UI load times, smaller pkg sizes etc).

Will Realm ever be in a position to support Hermes on React Native.

It’s not clear to me from your answer. I’m sensing  that you’re saying that Hermes and Realm will never be a perfect fit.

**Comment by HSReact** [comment]
@martinpoulsen my question exactly, I’m not getting a clear picture

**Comment by takameyer** [comment]
> Will Realm ever be in a position to support Hermes on React Native.

Yes, you can do this today.  `npm install realm@11.0.0-rc.1` with React Native `0.69.*`

**Comment by mordechaim** [comment]
> > Currently our focus is on JSI. TurboModules will be worth investigating at some point.
> 
> Does this mean that enabling the new architecture will not be an option in the foreseeable future when depending on Realm?

The new architecture was designed for gradual migration. You can enable the new architecture and still have libraries that use legacy modules.

The JSI is available regardless.

**Comment by tedi** [comment]
We've been running realm with hermes in production on IOS the past 4 months with no real issues. We recently ran into some issues when upgrading to RN .69x but had to revert back to .68 for a few other reasons. 

We have kept a close eye on local db related issues since we updated to the hermes supported realm and are happy to report everything has been working great on our end. We've been using it for fairly heavy operations, often managing thousands of records at a time and haven't run into any problems at all. 

Thank you for your work Realm team.

**Comment by tomduncalf** [comment]
**TL;DR**: Update on `v11` performance regression. With RN v0.70, we are seeing benchmark results around 0.6x vs `master` with Hermes enabled, and 0.4x vs `master` with Hermes disabled. Note that this is a reasonable worst case, and for many use cases we would not expect the regression to have an impact. We expect RN v0.71 should allow us to greatly improve performance for Hermes users, but unfortunately there is nothing we can do for users with Hermes disabled without changes to RN, so our recommendation to these users is to move to using Hermes or to stay on `v10` if your use case is performance sensitive.

—

Hi all,

I wanted to provide a bit more detail on the situation related to performance regressions referenced above and in https://github.com/realm/realm-js/issues/4443, as I've been doing a lot of research into this.

To briefly summarise, in order to support Hermes we had to move from interacting with the JavaScriptCore engine (JSC) directly, to using JSI (which is an abstraction layer for working with Hermes and JSC). This meant we had to change the way we attach C++ objects to JS objects, as JSI does not support the mechanism we used previously (JSC's [JSObjectSetPrivate](https://developer.apple.com/documentation/javascriptcore/1451626-jsobjectsetprivate)). 

Unfortunately, the best method we were able to find to do this with JSI is quite a lot slower, and this results in performance regressions, as accessing the C++ object is a "hot path" in Realm. I should stress that this should only be an issue in “worst case” scenarios, e.g. inserting many objects in a loop, and for most common use cases, we wouldn't expect there to be any noticeable regression.

The current situation on our `v11` branch with the final React Native 0.70 release is that our benchmark (which repeatedly accesses an object in a loop as fast as it can) performs around 0.6x the speed of `master` with Hermes enabled (this is a vast improvement on 0.69, where it was around 0.25x), and around 0.4x with Hermes disabled.

We've been working with Meta to try to understand the options available to us to improve this situation, and the good news is that in React Native v0.71, a new API is being introduced for attaching C++ objects to JS objects, and our initial testing suggests that this should bring performance much closer to parity with `master` when using the Hermes engine (we have a [draft PR](https://github.com/realm/realm-js/pull/4871) to add support for this when RN v0.71 is ready).

Unfortunately, this API is not currently available when Hermes is disabled, and so we do not expect RN v0.71 to improve performance for users with Hermes disabled. We have created a [proposal to implement this support in React Native](https://github.com/react-native-community/discussions-and-proposals/issues/505). Until that happens (which cannot be guaranteed, as I imagine Meta are mainly focussed on Hermes), we would expect performance using `v11` **without Hermes enabled** to suffer from this regression. Whether this regression is significant or not would depend on your workload. 

We’ve exhausted every other avenue we can think of in trying to find a solution for this, but without support for it at the React Native level, there’s no way we can work around this (while still supporting libraries such as Reanimated), so our recommendation for users who are not using Hermes and for whom performance is critical is to migrate to using Hermes (which is now the default as of React Native v0.70), or to remain on Realm JS `v10`.

Thanks,
Tom

**Comment by joaotmachado** [comment]
Have there been any changes to the RC in the last few days? 

The same projects with RN 0.70.1 + Realm 11.0.0-rc.2 + Hermes true are slow again. About 4 days ago it was ok.

Both Android and iOS.

**Comment by kneth** [comment]
@joaotmachado We don't modify versions after release. Can it be explained by a different use pattern? If so, we are interesting to hear about it as it might help us to optimise in the future.

**Comment by kneth** [comment]
Yesterday we released [version 11.0.0](https://github.com/realm/realm-js/releases/tag/v11.0.0) and we now consider Hermes a fully supported JavaScript engine. 

I would like to thank everybody who have tried our prereleases, provided feedback and showed us patience and trust during the integration with Hermes. It has been a large effort for a small team, and we have been motivated by your support.

**Comment by HSReact** [comment]
Great news and thank you all for your hard work and in exercising patience whilst accepting our feedback.

**Comment by 8BallBomBom** [comment]
Been a long road, thanks for the great work 👍🏻
